File: app/src/main/java/com/example/movieslisttask/view/ui/fragments/MoviesFragment.java
package com.example.movieslisttask.view.ui.fragments;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.databinding.DataBindingUtil;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProviders;
import androidx.recyclerview.widget.DefaultItemAnimator;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.example.movieslisttask.R;
import com.example.movieslisttask.databinding.FragmentMoviesBinding;
import com.example.movieslisttask.model.Movie;
import com.example.movieslisttask.tools.Consts;
import com.example.movieslisttask.view.adapter.MovieAdapter;
import com.example.movieslisttask.viewmodel.MoviesViewModel;

import java.util.ArrayList;

/**
 * Created by alexkorolov on 2020-01-17.
 */
public class MoviesFragment extends Fragment {

    private ArrayList<Movie> movies;
    private FragmentMoviesBinding moviesBinding;
    private MovieAdapter movieAdapter;
    private MoviesViewModel moviesViewModel;
    private boolean isFavoritesFrag = false;

    public static MoviesFragment newInstance(boolean favorites) {
        MoviesFragment retFrag = new MoviesFragment();

        Bundle args = new Bundle();
        args.putBoolean(Consts.SHOW_FAVORITES, favorites);
        retFrag.setArguments(args);

        return retFrag;
    }

    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {

        moviesViewModel = ViewModelProviders.of(this).get(MoviesViewModel.class);
        moviesBinding = DataBindingUtil.inflate(inflater, R.layout.fragment_movies, container, false);

        if (getArguments() != null) {
            isFavoritesFrag = getArguments().getBoolean(Consts.SHOW_FAVORITES);
        }

        getMovies();

        return moviesBinding.getRoot();
    }

    private void getMovies() {
        if (isFavoritesFrag) {
            moviesViewModel.getAllFavorites().observe(getViewLifecycleOwner(), moviesFromLiveData -> {
                movies = (ArrayList<Movie>) moviesFromLiveData;
                handleDataReceived();
            });
        } else {
            moviesViewModel.getallMovies().observe(getViewLifecycleOwner(), moviesFromLiveData -> {
                movies = (ArrayList<Movie>) moviesFromLiveData;
                handleDataReceived();
            });
        }
    }

    private void handleDataReceived() {
        if (movieAdapter == null) {
            RecyclerView recyclerView = moviesBinding.moviesList;
            movieAdapter = new MovieAdapter(getActivity(), movies);
            recyclerView.setLayoutManager(new LinearLayoutManager(getActivity()));
            recyclerView.setItemAnimator(new DefaultItemAnimator());
            recyclerView.setAdapter(movieAdapter);
        } else {
            movieAdapter.updateMoviesList(movies);
        }
    }
}

File: app/src/main/java/com/example/movieslisttask/ui/movies_list/MoviesFragment.kt
package com.example.movieslisttask.ui.movies_list

import android.app.Activity
import android.app.ActivityOptions
import android.content.Intent
import android.os.Bundle
import android.util.Pair
import android.view.View
import android.widget.ImageView
import androidx.fragment.app.Fragment
import androidx.lifecycle.Observer
import com.example.movieslisttask.R
import com.example.movieslisttask.db.entities.Movie
import com.example.movieslisttask.tools.Consts
import com.example.movieslisttask.ui.activities.MovieInfoActivity
import kotlinx.android.synthetic.main.fragment_movies.*
import org.koin.androidx.viewmodel.ext.android.viewModel


/**
 * Created by alexkorolov on 2020-01-17.
 */
class MoviesFragment : Fragment(R.layout.fragment_movies), MoviesAdapter.OnMovieClicked {

    private val moviesViewModel: MoviesViewModel by viewModel()
    private var isFavoritesFrag = false
    private var movieAdapter: MoviesAdapter = MoviesAdapter(listOf(), this)

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        isFavoritesFrag = arguments?.getBoolean(Consts.SHOW_FAVORITES) ?: false
        moviesList.adapter = movieAdapter
        getMovies()
    }

    private fun getMovies() {
        moviesViewModel.getMovies(isFavoritesFrag).observe(viewLifecycleOwner, Observer { moviesList ->
            movieAdapter.setData(moviesList)
        })
    }

    override fun addToFavorite(movie: Movie) {
        moviesViewModel.handleFavoriteMovie(movie)
    }

    override fun openFullMovie(movie: Movie, image: ImageView) {
        val intent = Intent(activity, MovieInfoActivity::class.java)
        intent.putExtra("movie", movie)

        val imagePair = Pair<View, String>(image, "image_transition")
        val options = ActivityOptions.makeSceneTransitionAnimation(context as Activity?, imagePair)

        startActivity(intent, options.toBundle())
    }

    companion object {
        fun newInstance(favorites: Boolean) = MoviesFragment().apply {
            arguments = Bundle().apply {
                putBoolean(Consts.SHOW_FAVORITES, favorites)
            }
        }
    }
}